fili-core/src/main/java/com/yahoo/bard/webservice/data/dimension/AvroDimensionRowParser.java

```diff
@@ -0,0 +1,185 @@
+
+        // Returns true, only if all the dimension fields present in the configured dimension exist in the AVRO schema
+        return convertedDimensionFields.values().containsAll(avroFields);
+    }
+
+    /**
+     * Parses the avro file and populates the dimension rows after validating the schema.
+     *
+     * @param dimension The dimension object used to configure the dimension
+     * @param avroFilePath The path of the AVRO data file (.avro)
+     * @param avroSchemaPath The path of the AVRO schema file (.avsc)
+     *
+     * @throws NullPointerException thrown if JSON object `fields` is not present
+     * @throws IOException thrown if there is error parsing the avro files
+     */
+    public void parseAvroFileDimensionRows(Dimension dimension, String avroFilePath, String avroSchemaPath)
+        throws NullPointerException, IOException {
+
+        if (isSchemaValid(dimension, avroSchemaPath)) {
+
+            JsonNode jsonAvroSchema = convertFileToJsonNode(avroSchemaPath);
+
+            // Creates an AVRO schema object based on the parsed AVRO schema file (.avsc)
+            Schema schema = new Schema.Parser().parse(jsonAvroSchema.toString());
+
+            // Creates an AVRO DatumReader object based on the AVRO schema object
+            DatumReader<GenericRecord> datumReader = new GenericDatumReader<>(schema);
+
+            // Creates an AVRO DataFileReader object that reads the AVRO data file one record at a time
+            DataFileReader<GenericRecord> dataFileReader =
+                new DataFileReader<>(new File(avroFilePath), datumReader);
+
+            // Represents each record in the avro data file
+            GenericRecord avroData = null;
+
+            while (dataFileReader.hasNext()) {
+                avroData = dataFileReader.next(avroData);
+
+                // Creates a JsonObject from the AVRO GenericRecord object
+                JsonNode avroDataJson = MAPPER.readTree(new JsonFactory().createParser(avroData.toString()));
+
+                // Constructs a Map of the converted dimension field name and the corresponding value in the avro file
+                Map<String, String> dimensionFieldKeyValues = dimension.getDimensionFields()
+                        .stream()
+                        .collect(Collectors.toMap(dimensionField -> dimensionField.getName(),
+                                                  // Looks up converted field name from the `convertedDimensionFields`
+                                                  // Parsed as text to avoid the extra quotes around string objects
+                                                  dimensionField -> avroDataJson
+                                                        .get(convertedDimensionFields
+                                                                .get(dimensionField.getName())
+                                                        ).asText()
+                                                )
```

Comment:
@michael-mclawhorn the challenge here is that "pull the info off of the avro record" is pretty much all this is doing. The only other task in here is to do that for each field on the dimension (so, another loop). Another way of looking at it would be:

```java
StreamSupport.stream(dataFileReader.spliterator(), false)
        .map(genericRecord -> dimension.getDimensionFields().stream().collect(
                        Collectors.toMap(
                                DimensionField::getName,
                                dimensionField -> genericRecord.get(
                                        dimensionFieldMapper.converter(dimension, dimensionField)
                                ).toString()
                        )
                )
        )
        .map(dimension::parseDimensionRow)
        .collect(Collectors.toSet());
```

One thing that _could_ be reduced (which the current code does via the shared data structure) would be to memoize the field conversions. (see my other comment)